### PR TITLE
mac-capture: Improve window capture performance

### PR DESCRIPTION
**Pull request by gxalpha**
### Description

Replaces the current implementation of Window Capture, which uses Core
Graphics to render a bitmap raster, with one that accesses the window's
CGImage bitmap directly, blits it, and hands that off as a frame to OBS
instead.

Taken out of #5680.

### Motivation and Context

@dossy did great work figuring this out, the entire pull request
is just rather big (also containing other changes and cleanups),
with everything being in one commit which makes it somewhat
difficult to review (It also contains changes to how the window
selection works, which is somewhat controversial and will require
more debate).
The performance enhancement itself though is pretty crucial for
anyone using window capture, it should not get held up in other
debates.

### How Has This Been Tested?

macOS 12.4 Beta
I can no longer count individual frames when using window capture :P
No seriously, while I haven't done any benchmarking (others have
done that on the original PR already) it's *noticeably* less laggy (on
my rather old machine).

### Types of changes

- Performance enhancement (non-breaking change which improves efficiency)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: plugins/mac-capture/mac-window-capture.m [hunk 1]
```diff
@@ -50,28 +50,28 @@ static inline void capture_frame(struct window_capture *wc)
...
-	uint8_t *data = wc->buffer.array;
-
-	CGContextRef cg_context = CGBitmapContextCreate(
-		data, width, height, 8, width * 4, wc->color_space,
-		kCGBitmapByteOrder32Host | kCGImageAlphaPremultipliedFirst);
-	CGContextSetBlendMode(cg_context, kCGBlendModeCopy);
-	CGContextDrawImage(cg_context, rect, img);
-	CGContextRelease(cg_context);
-	CGImageRelease(img);
+	if ((!width && !height) || CGImageGetBitsPerPixel(img) != 32 ||
```
Comment (PatTheMav) [inline plugins/mac-capture/mac-window-capture.m:53]: If I read this right, if either width or height is missing (but not both), the function won't return early.

File: plugins/mac-capture/mac-window-capture.m [hunk 2]
```diff
@@ -50,28 +50,28 @@ static inline void capture_frame(struct window_capture *wc)
...
 		.width = width,
 		.height = height,
-		.data[0] = data,
-		.linesize[0] = width * 4,
+		.data[0] = (uint8_t *)CFDataGetBytePtr(data),
+		.linesize[0] = CGImageGetBytesPerRow(img),
 		.timestamp = ts,
 	};
 
 	obs_source_output_video(wc->source, &frame);
```
Comment (PatTheMav) [inline plugins/mac-capture/mac-window-capture.m:71]: I take it `obs_source_output_video` blocks and uploads the data to the GPU before returning?

Comment (gxalpha) [reply plugins/mac-capture/mac-window-capture.m:71]: I'm not entirely sure about the internal details, but I do know that it does a copy of the memory before returning, so we can release the data immediately afterwards (if that's the question).

File: plugins/mac-capture/mac-window-capture.m (hunk 1 above)
Comment (gxalpha) [reply plugins/mac-capture/mac-window-capture.m:53]: You are right, that should probably be an "or".

File: plugins/mac-capture/mac-window-capture.m (hunk 2 above)
Comment (PatTheMav) [reply plugins/mac-capture/mac-window-capture.m:71]: Yeah should be fine, there is a `memcpy` involved.

**Pull request merged**